tendermint/src/block/commit_sig.rs

```diff
@@ -1,78 +1,182 @@
+            // Implements decision: https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-025-commit.md#decision
+            fn visit_map<A>(self, map: A) -> Result<Self::Value, <A as MapAccess<'de>>::Error>
+            where
+                A: MapAccess<'de>,
+            {
+                // Instead of manually deserializing the whole struct (cumbersome), we use annotations
+                fn option_signature<'de, D>(deserializer: D) -> Result<Option<Signature>, D::Error>
+                where
+                    D: Deserializer<'de>,
+                {
+                    Deserialize::deserialize(deserializer).map(|x: Option<_>| x.unwrap_or(None))
+                }
+                #[derive(Deserialize)]
+                struct CommitSigSpec {
+                    block_id_flag: u8,
+                    validator_address: account::Id,
+                    #[serde(default)]
+                    timestamp: Option<Time>,
+                    #[serde(default, deserialize_with = "option_signature")]
+                    signature: Option<Signature>,
+                }
 
-    /// Signature
-    #[serde(deserialize_with = "serializers::parse_non_empty_signature")]
-    pub signature: Option<Signature>,
-}
+                // `MapAccessDeserializer` is a wrapper that turns a `MapAccess`
+                // into a `Deserializer`, allowing it to be used as the input to T's
+                // `Deserialize` implementation. T then deserializes itself using
+                // the entries from the map visitor.
+                let incoming =
+                    CommitSigSpec::deserialize(de::value::MapAccessDeserializer::new(map))?;
 
-impl CommitSig {
-    /// Checks if a validator's vote is absent
-    pub fn is_absent(&self) -> bool {
-        self.block_id_flag == BlockIDFlag::BlockIDFlagAbsent
+                // Validate CommitSig (strict)
+                match incoming.block_id_flag {
+                    // BlockIDFlagAbsent
+                    1 => {
+                        if incoming.timestamp.is_some() {
+                            Err(A::Error::custom(format!(
+                                "timestamp is present for BlockIDFlagAbsent CommitSig {}",
+                                incoming.timestamp.unwrap()
+                            )))
+                        } else {
+                            if incoming.signature.is_some() {
+                                Err(A::Error::custom(format!(
+                                    "signature is present for BlockIDFlagAbsent CommitSig {:?}",
+                                    incoming.signature.unwrap()
+                                )))
+                            } else {
+                                Ok(CommitSig::BlockIDFlagAbsent {
+                                    validator_address: incoming.validator_address,
+                                })
+                            }
+                        }
```

Comment:
Could also be written as:
```suggestion
                       if incoming.timestamp.is_some() {
                            Err(A::Error::custom(format!(
                                "timestamp is present for BlockIDFlagAbsent CommitSig {}",
                                incoming.timestamp.unwrap()
                            )))
                        }
                        if incoming.signature.is_some() {
                            Err(A::Error::custom(format!(
                                "signature is present for BlockIDFlagAbsent CommitSig {:?}",
                                incoming.signature.unwrap()
                            )))
                        }
                        Ok(CommitSig::BlockIDFlagAbsent {
                            validator_address: incoming.validator_address,
                        })
```

Reply:
If Clippy says so, it needs to happen :)